shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+        }
+
+        // enqueue a sync + data push
+        this.debounced('record', 1000, () => {
+            this.syncToStorage()
+            this.sendToNewTab('following a recorded tracker-blocked event')
+        })
+    }
+
+    /**
+     * Persist data into the extensions storage in the following format
+     */
+    syncToStorage () {
+        const serializedData = this.stats.serialize()
+        const toSync = {
+            [NewTabTrackerStats.storageKey]: {
+                // making this an object to ensure we can store more things under this
+                // namespace later if we need to
+                stats: serializedData
+            }
+        }
+        browserWrapper.syncToStorage(toSync)
+    }
+
+    /**
+     * Attempt to re-populate stats from storage.
+     * @returns {Promise<void>}
+     */
+    async restoreFromStorage () {
+        try {
+            const prev = await browserWrapper.getFromStorage(NewTabTrackerStats.storageKey)
+            if (prev) {
+                this.stats.deserialize(prev.stats)
+            }
+        } catch (e) {
+            console.warn("could not deserialize data from _cachedDisplayData 'trackerStats' storage")
+        }
+
+        // also evictExpired once we've restored
+        this.stats.evictExpired()
+    }
+
+    /**
+     * A single handler for all incoming events relating
+     * to the new tab page such as heartbeat, sendInitial etc
+     * @returns {void}
+     */
+    handleIncomingEvent (event) {
+        /**
+         * Handle every message prefixed with `newTabPage_`
+         * For now every event just causes new data to be pushed
+         */
+        if (typeof event.messageType === 'string' && event.messageType.startsWith(NewTabTrackerStats.eventPrefix)) {
+            if (event.messageType in incoming) {
+                this.sendToNewTab(`response to '${event.messageType}'`)
+            } else {
+                console.error('unhandled event prefixed with: ', NewTabTrackerStats.eventPrefix, event)
+            }
+        }
+    }
+
+    /**
+     * Convert locally stored data into data that can be consumed by a UI
+     * @param {string} reason - a reason or path that caused this
+     */
+    sendToNewTab (reason) {
+        if (!reason) throw new Error("you must provide a 'reason' for sending new data")
+        this.debounced('sendToNewTab', 200, () => this._publish(reason))
+    }
+
+    /**
+     * Private method for doing the actual send in a single place
+     * that's easy to debug
+     */
+    _publish (reason = 'unknown') {
+        if (this._debug) {
+            console.info(`sending new tab data because: ${reason}`)
+        }
+        /** @type {import("zod").infer<typeof import("../newtab/schema").dataMessage>} */
+        const msg = {
+            messageType: outgoing.newTabPage_data,
+            options: this.toDisplayData()
+        }
+        chrome.runtime.sendMessage(msg)
```

Comment:
We should check other places that will receive this message and make sure they discard it correctly. I see, for example, [this code](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/content-scripts/content-scope-messaging.js#L41-L45) which looks like it might forward these messages blindly.